group/assignment5/Assignment5.java

```diff
@@ -0,0 +1,149 @@
+        LinkedList<LinkedList<Character>> allPossibleAlphabets = new LinkedList<>();
+
+        if (dictionaryList.isEmpty()) {
+            return allPossibleAlphabets;
+        }
+
+        if (dictionaryList.size() == 1) {
+            formTopologicalGraphFromOneWordDictionary(dictionaryList.get(0));
+        } else {
+            formTopologicalGraph(dictionaryList);
+        }
+
+        allPossibleAlphabets = graphOfLetters.allTopologicalSortOrders();
+
+        return allPossibleAlphabets;
+    }
+
+    public void formTopologicalGraphFromOneWordDictionary(String word) {
+        for (int i = 0; i < word.length(); i++) {
+            addOrFetchVertex(word.charAt(i));
+        }
+    }
+
+    public void formTopologicalGraph(List<String> dictionaryList) {
+        for (int i = 0; i < dictionaryList.size() - 1; i++) {
+            String firstWord = dictionaryList.get(i);
+            String secondWord = dictionaryList.get(i + 1);
+
+            int index = 0;
+            int firstWordLength = firstWord.length();
+            int secondWordLength = secondWord.length();
+
+            boolean checkForGraphEdges = true;
+            while (index < firstWordLength && index < secondWordLength) {
+                Character firstWordsCharacter = firstWord.charAt(index);
+                Character secondWordsCharacter = secondWord.charAt(index);
+
+                // insert firstWord's character if it isn't in the map and make a space for it in the graph; or just fetch it from map if it exists
+                Vertex firstWordsCharVertex = addOrFetchVertex(firstWordsCharacter);
+                //insert secondWrod's character if it isn't in the map (also works if the characters are the same, because it has just been inserted); or just fetch it from map if it exists
+                Vertex secondWordsCharVertex = addOrFetchVertex(secondWordsCharacter);
+
+                if (!firstWordsCharacter.equals(secondWordsCharacter) && checkForGraphEdges) {
+                    // different characters, both inserted in the map if needed
+                    // if they are the first two different characters discovered, 
+                    // insert edge in graph from firstWordsCharacter to secondWordsCharacter
+                    checkForGraphEdges = false;
+                    graphOfLetters.addEdge(firstWordsCharVertex, secondWordsCharVertex);
+                }
+                index++;
+            }
+
+            int secondWordIndex = index;
+
+            // process until the end of the first word if it is longer
+            while (index < firstWordLength) {
+                Character firstWordsCharacter = firstWord.charAt(index);
+                addOrFetchVertex(firstWordsCharacter);
+                index++;
+            }
+
+            // need to process the last word in dictionary, if it is longer
+            if (isSecondWordLastInDictionary(dictionaryList.size(), i + 1)) {

```

Comment:
Checking if it's the last word is a good idea, I've realised I do such a processing for every word apart from the last one twice.